java/code/src/com/redhat/rhn/manager/system/SystemManager.java

```diff
@@ -3775,4 +3781,78 @@ else if (pkgState == PackageStates.PURGED) { // using PURGED flag for unmanaged
         StateFactory.save(stateRev);
         StatesAPI.generateServerPackageState(minion);
     }
+
+    /**
+     * Update MgrServerInfo with current grains data
+     *
+     * @param minion the minion which is a Mgr Server
+     * @param grains grains from the minion
+     */
+    public static void updateMgrServerInfo(MinionServer minion, ValueMap grains) {
+        // Check for Uyuni Server and create basic info
+        if (grains.getOptionalAsBoolean("is_mgr_server").orElse(false)) {
+            MgrServerInfo serverInfo = Optional.ofNullable(minion.getMgrServerInfo()).orElse(new MgrServerInfo());
+            String oldHost = serverInfo.getReportDbHost();
+            String oldName = serverInfo.getReportDbName();
+
+            serverInfo.setVersion(PackageEvrFactory.lookupOrCreatePackageEvr(null,
+                    grains.getOptionalAsString("version").orElse("0"),
+                    "1", minion.getPackageType()));
+            serverInfo.setReportDbName(grains.getValueAsString("report_db_name"));
+            serverInfo.setReportDbHost(grains.getValueAsString("report_db_host"));
+            serverInfo.setReportDbPort((grains.getValueAsLong("report_db_port").orElse(5432L)).intValue());
+            serverInfo.setServer(minion);
+            minion.setMgrServerInfo(serverInfo);
+
+            if (!StringUtils.isAnyBlank(oldHost, oldName) &&
+                    !(oldHost.equals(serverInfo.getReportDbHost()) &&
+                            oldName.equals(serverInfo.getReportDbName()))) {
+                // something changed, we better reset the user
+                setReportDbUser(minion, false);
+            }
+        }
+        else {
+            ServerFactory.dropMgrServerInfo(minion);
+            // Should we try to drop the credentials on the reportdb?
+        }
+    }
+
+    /**
+     * Set the User and Password for the report database in MgrServerInfo.
+     * It trigger also a state apply to set this user in the report database.
+     *
+     * @param minion the Mgr Server
+     * @param forcePwChange force a password change
+     */
+    public static void setReportDbUser(MinionServer minion, boolean forcePwChange) {
+        // Create a report db user when system is a mgr server
+        if (minion.isMgrServer()) {
```

Comment:
Also here an early out would avoid the nesting and improve readability

Reply:
changed